cpp/src/fil/infer.cu

```diff
@@ -160,6 +198,130 @@ struct tree_aggregator_t {
   }
 };
 
+template <int NITEMS>
+struct tree_aggregator_t<NITEMS, TREE_PER_CLASS_FEW_CLASSES> {
+  vec<NITEMS, float> acc;
+  void* tmp_storage;
+  int num_classes;
+
+  static size_t smem_finalize_footprint(int num_classes) {
+    size_t phase1 =
+      (FIL_TPB - FIL_TPB % num_classes) * sizeof(vec<NITEMS, float>);
+    size_t phase2 =
+      sizeof(typename BlockReduceHostMultiClass<NITEMS>::TempStorage);
+    return std::max(phase1, phase2);
+  }
+
+  static size_t smem_accumulate_footprint(int num_classes) { return 0; }
+
+  __device__ __forceinline__ tree_aggregator_t(int num_classes_,
+                                               void* shared_workspace, size_t)
+    : tmp_storage(shared_workspace), num_classes(num_classes_) {}
+
+  __device__ __forceinline__ void accumulate(
+    vec<NITEMS, float> single_tree_prediction, int tree) {
+    acc += single_tree_prediction;
+  }
+
+  __device__ __forceinline__ void finalize(float* out, int num_rows,
+                                           int num_outputs) {
+    __syncthreads();  // free up input row
+    // load margin into shared memory
+    best_margin_label<NITEMS> best;
+    vec<NITEMS, float>* per_thread = (vec<NITEMS, float>*)tmp_storage;
+    if (threadIdx.x >= num_classes) per_thread[threadIdx.x] = acc;
+
+    __syncthreads();
+    // reduce per-thread margin summand into per-class complete margin
+    // (for each of the NITEMS rows)
+    if (threadIdx.x < num_classes) {
```

Comment:
Add a TODO to use tree reduction to improve performance when `num_classes` is small.